src/lib/parsers/sol-json.c

```diff
@@ -567,3 +567,144 @@ sol_json_is_valid_type(struct sol_json_scanner *scanner, enum sol_json_type type
            sol_json_scanner_skip_over(scanner, &token) &&
            token.end == last_position;
 }
+
+SOL_API int
+sol_json_token_get_unescaped_string(const struct sol_json_token *token, struct sol_buffer *buffer)
+{
+    int r;
+    const char *start, *p;
+    bool is_escaped = false;
+    char new_char;
+
+    SOL_NULL_CHECK(buffer, -EINVAL);
+    sol_buffer_init_flags(buffer, NULL, 0, SOL_BUFFER_FLAGS_NO_NUL_BYTE);
+
+    SOL_NULL_CHECK(token, -EINVAL);
+    SOL_NULL_CHECK(token->start, -EINVAL);
+    SOL_NULL_CHECK(token->end, -EINVAL);
+
+    if (sol_json_token_get_type(token) != SOL_JSON_TYPE_STRING) {
+        sol_buffer_init_flags(buffer, (char *)token->start,
+            sol_json_token_get_size(token), SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED);
+        buffer->used = buffer->capacity;
+        return 0;
+    }
+
+    if (*(token->end - 1) != '"')
+        goto invalid_json_string;
+
+    for (start = p = token->start + 1; p < token->end - 1; p++) {
+        if (!is_escaped && *p == '\\') {
+            struct sol_str_slice slice = { .data = start, .len = p - start };
+
+            r = sol_buffer_append_slice(buffer, slice);
+            SOL_INT_CHECK_GOTO(r, < 0, error);
+            is_escaped = true;
+        } else if (is_escaped) {
+            is_escaped = false;
+            start = p + 1;
+            switch (*p) {
+            case '\\':
+                new_char = '\\';
+                break;
+            case '/':
+                new_char = '/';
+                break;
+            case '"':
+                new_char = '"';
+                break;
+            case 'b':
+                new_char = '\b';
+                break;
+            case 'r':
+                new_char = '\r';
+                break;
+            case 'n':
+                new_char = '\n';
+                break;
+            case 'f':
+                new_char = '\f';
+                break;
+            case 't':
+                new_char = '\t';
+                break;
+            case 'u':
+                if (p + 4 < token->end - 1) {
+                    char hexdigits[3], *endptr;
+                    uint16_t n;
+                    hexdigits[2] = '\0';
+
+                    //Add first unicode byte
+                    memcpy(hexdigits, p + 3, 2);
+                    n = strtoul(hexdigits, &endptr, 16);
+                    SOL_INT_CHECK(errno, < 0, -errno);
```

Comment:
you must manually set `errno = 0` before calling `strtoul()`, the man page says:

```
       Since strtoul() can legitimately return 0 or ULONG_MAX (ULLONG_MAX  for
       strtoull()) on both success and failure, the calling program should set
       errno to 0 before the call, and then determine if an error occurred  by
       checking whether errno has a nonzero value after the call.
```